subsys/bluetooth/host/gatt.c

```diff
@@ -2421,14 +2439,16 @@ static uint8_t notify_cb(const struct bt_gatt_attr *attr, uint16_t handle,
 		 * Client Characteristic Configuration descriptor may occur
 		 * in any position within the characteristic definition after
 		 * the Characteristic Value.
+		 * Only notify or indicate devices which are subscribed.
 		 */
-		if (data->type == BT_GATT_CCC_INDICATE) {
-			err = gatt_indicate(conn, data->handle,
-					    data->ind_params);
+		if ((data->type == BT_GATT_CCC_INDICATE) &&
+		    (cfg->value & BT_GATT_CCC_INDICATE)) {
+			err = gatt_indicate(conn, data->handle, data->ind_params);
 			if (err == 0) {
 				data->ind_params->_ref++;
 			}
-		} else {
+		} else if ((data->type == BT_GATT_CCC_NOTIFY) &&
+			   (cfg->value & BT_GATT_CCC_NOTIFY)) {
 			err = gatt_notify(conn, data->handle, data->nfy_params);
 		}
```

Comment:
In what kind of scenario would neither of these branches be entered, is that normal/valid? (if not, then maybe an else-branch would be good with a warning and `err = 0` to avoid the initialisation upon declaration of the variable.

Btw (since GitHub doesn't let me comment on unchanged lines), there seems to be a missing `bt_conn_unref()` for the "Link is not encrypted" branch a bit earlier in this function. Might want to fix that in a separate commit.

Reply:
I think this could be valid if the user doesn't use the zephyr macros to define their characteristics. Here it is taken into account, but there are other places where it definitely is not.
I don't really know what to do about it, should I fix it here, or make an issue to fix it in the whole file in a later PR.

Added the missing unref in a separate commit.